Feat: Add Jira Receiver to Alertmanager Config CR

nutmos opened the pull request:
## Description

Add a new JIRA integration for alertmanager.

Relate #7285
Co-author @dongjiang1989 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
E2E Test Cases

## Changelog entry

```release-note
- Add a new JIRA integration for Alertmanager in the custom resource
```

dongjiang1989 committed:
support jira Receiver for alertmanager

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
modify spelling by codereview

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
update

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'main' into support-alertmanager-jira

dongjiang1989 committed:
fix merge conflict

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Update pkg/alertmanager/amcfg.go

Co-authored-by: Nicolas Takashi <nicolas.tcs@hotmail.com>

dongjiang1989 committed:
fix issue_type mandatory field

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
fix unittest case

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Merge branch 'main' into support-alertmanager-jira

dongjiang1989 committed:
make generate

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
fix golangci lint

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
fix make generate

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Merge branch 'main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
fix ConfigBuilder

Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

dongjiang1989 committed:
Merge branch 'prometheus-operator:main' into support-alertmanager-jira

nutmos committed:
fix merge conflict and update doc

nutmos committed:
update generated files

nutmos committed:
update test case and var name

nutmos committed:
fix typo

nutmos committed:
update generated files

nutmos committed:
update description

nutmos committed:
update generated files

nutmos committed:
update fix linter

nutmos committed:
update generated files

nutmos committed:
add optional

nutmos committed:
update quote

nutmos committed:
fix sanitize func

nutmos committed:
update test case name

nutmos committed:
update test case

nutmos committed:
move validation method up

nutmos committed:
update description

nutmos committed:
move validate method up

nutmos committed:
trigger ci

heliapb commented on pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:
```diff
@@ -1132,6 +1136,137 @@ type MSTeamsConfig struct {
+	WontFixResolution *string `json:"wontFixResolution,omitempty"`
+
+	// reopenDuration defines to reopen the issue when it is not older than this value (rounded down to the nearest minute).
+	// The 'resolutiondate' field in Jira is used to determine the age of the issue.
```
```suggestion
	// The 'resolution date' field in Jira is used to determine the age of the issue.
```
Hi seems there new spell check workflow got this one.

nutmos replied:
Done, please have a look.

nutmos committed:
add resolutiondate to dictionary

nutmos committed:
Merge branch 'main' into support-alertmanager-jira

nutmos committed:
merge main and update generated files

nutmos committed:
update issue type and project check

nutmos committed:
print error

nutmos committed:
update print

nutmos committed:
update error

nutmos committed:
add error printing

nutmos committed:
update test case

nutmos committed:
add api url in the test cases

nutmos committed:
update api url test case

nutmos committed:
remove unused import

nutmos committed:
trigger ci

nutmos committed:
add api type field

nutmos committed:
add api type in jira config and parser

cjedro commented:
@nutmos the unit tests are failing now after your change

```
-- FAIL: TestSanitizeJiraConfig (0.00s)
    --- PASS: TestSanitizeJiraConfig/jira_configs_returns_error_for_unsupported_versions (0.00s)
    --- PASS: TestSanitizeJiraConfig/jira_configs_allows_for_supported_versions (0.00s)
    --- PASS: TestSanitizeJiraConfig/jira_configs_returns_error_for_missing_mandatory_fields (0.00s)
    --- PASS: TestSanitizeJiraConfig/jira_configs_with_send_resolved (0.00s)
    --- FAIL: TestSanitizeJiraConfig/jira_configs_with_api_type (0.00s)
    --- FAIL: TestSanitizeJiraConfig/jira_configs_with_incorrect_api_type (0.00s)
    --- FAIL: TestSanitizeJiraConfig/jira_configs_with_api_type_unsupported_version (0.00s)
```

nutmos committed:
update test cases

nutmos committed:
update golden file

nutmos committed:
trigger ci

nutmos commented:
@heliapb Could you please help me review?

nutmos commented:
@heliapb Any chance for reviewing this PR?

nutmos commented:
@heliapb Are you able to have a review please?

nutmos committed:
update validation/test case typo

nutmos commented on pkg/alertmanager/validation/v1beta1/validation_test.go:
```diff
@@ -530,6 +551,28 @@ func TestValidateAlertmanagerConfig(t *testing.T) {
+							JiraConfigs: []monitoringv1beta1.JiraConfig{
+								{
+									Project: "projectA",
+									APIURL:  ptr.To("http://test.com"),
```
Updated

nutmos committed:
update config

nutmos committed:
update apiversion

nutmos committed:
update operator status

nutmos committed:
update test case: add required field

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
update suggested by linter

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
resolve merge conflict

nutmos committed:
trigger ci

nutmos committed:
trigger ci

nutmos committed:
resolve merge conflict

nutmos committed:
api: autogenerate file

resolve merge conflict

nutmos committed:
alertmanager api & cfg: update api validation & type

to cover more validation cases

nutmos committed:
amcfg: update test case

to reflect the new variable

nutmos committed:
e2e: add jira test case

to cover the real scenario

nutmos committed:
trigger ci

nutmos committed:
e2e: remove test case to prove concept

nutmos commented:
Wait for #8218 to merge.

nutmos committed:
alertmanager: resolve merge conflict

to resolve merge conflict

nutmos committed:
alertmanager: update test case

to resolve issue after merge

nutmos committed:
api: autogenerate file

to reflect the new changes

nutmos committed:
validation: update validation method and add test case for URL validation

to use the new method and case to cover more validations

nutmos committed:
validation: update validation method

to fix linter

nutmos committed:
amcfg: update test file by removing the log print

to remove unused code

nutmos committed:
api: add linter no validation

to skip bool validation

nutmos committed:
alertmanager: update test case

to change the CR parsing order

nutmos commented:
@simonpasquier @heliapb  Hi, could you please have a look at this as well? This is pending for a long time. Thank you!

nutmos committed:
amcfg: revert unrelated change

to prevent non-related change to happen with this pr

nutmos committed:
amcfg: revert unrelated change

to prevent non-related change to happen with this pr

simonpasquier committed:
fix: include operator.prometheus.io/version in full CRDs

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
test: add test for #8312

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
Add CreateStatefulSetOrPatchLabels() to k8sutil package

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
Use new CreateStatefulSetOrPatchLabels() function in controllers

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
Add end-to-end tests for Alertmanager and ThanosRuler

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

Yashika0724 committed:
fix: compute input hash before creating ThanosRuler StatefulSet

Signed-off-by: Yashika0724 <ssyashika1311@gmail.com>

nutmos committed:
resolve merge conflicts

rajatvig committed:
Merge pull request #8348 from rajatvig/feat/bodysizelimit-sc

scrapeconfig: add bodySizeLimit field

dependabot[bot] committed:
build(deps): bump github.com/prometheus/alertmanager

Bumps [github.com/prometheus/alertmanager](https://github.com/prometheus/alertmanager) from 0.30.1 to 0.31.0.
- [Release notes](https://github.com/prometheus/alertmanager/releases)
- [Changelog](https://github.com/prometheus/alertmanager/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prometheus/alertmanager/compare/v0.30.1...v0.31.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/alertmanager
  dependency-version: 0.31.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Arpit529Srivastava committed:
feat: add promtool validation for golden files (#8216)

* feat: add promtool validation for golden files

Signed-off-by: arpit529srivastava <arpitsrivastava529@gmail.com>

* fix: add goldenfiles and testdata to cspell dictionary

Signed-off-by: arpit529srivastava <arpitsrivastava529@gmail.com>

* fix: clarify note on excluded files in golden file validation

Signed-off-by: arpit529srivastava <arpitsrivastava529@gmail.com>

* refactor: use tools.go pattern for promtool

Signed-off-by: arpit529srivastava <arpitsrivastava529@gmail.com>

---------

Signed-off-by: arpit529srivastava <arpitsrivastava529@gmail.com>

heliapb committed:
feat: bump am to latest

Signed-off-by: Hélia Barroso <helia_barroso@hotmail.com>

heliapb committed:
feat: add httpheaders in am secret

Signed-off-by: Hélia Barroso <helia_barroso@hotmail.com>

simonpasquier committed:
chore: rephrase creation error messages

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

Arpit529Srivastava committed:
Merge pull request #8350 from Arpit529Srivastava/kubelet-cmd-flag

feat: add flag to disable insecure kubelet metrics port

nutmos committed:
Feat: add test cases for Alertmanager secret config Jira receiver (#8349)

* amcfg: add jira sanitize test cases for alertmanager secret config

* amcfg: add golden file

for the missing comparison

WHOIM1205 committed:
Merge pull request #8310 from WHOIM1205/fix-informer-nil-pointer-panic

Fix nil pointer panic in StatefulSet status reporting during informer cache races

Sanchit2662 committed:
fix: support IPv6 addresses in prober URL validation (#8354)

* fix: support IPv6 addresses in prober URL validation

Signed-off-by: Sanchit2662 <sanchit2662@gmail.com>

* docs: add comment explaining govalidator usage over stdlib

Signed-off-by: Sanchit2662 <sanchit2662@gmail.com>

---------

Signed-off-by: Sanchit2662 <sanchit2662@gmail.com>

kartikangiras committed:
fix: validate nonzero duration in AlertmanagerConfig (#8126)

* add valid alertmanagerconfig

* update validation check for global alertmanagerconfig object

* add validation marker

* add e2e test for amcfg

* fix depedency issues

* add sanitization logic & monitorv1duration field

* update tests with monitorv1duration field logic

* validate and update fields

* update import dependencies

* update unnecessary groupwait conversion

* update generated files and formatting

* update error handling && refine methods

* nonemptyduration validation for groupwait

* retrigger CI

* update validation for groupwait

* remove toplevel route validation for intervals

* simplify handling of zero durations

Zero durations are converted to empty strings which makes use of the
default value defined at a higher level in the routing tree.

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

* Fix generated file

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

* Fix test compilation

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

---------

Signed-off-by: Simon Pasquier <spasquie@redhat.com>
Co-authored-by: Simon Pasquier <spasquie@redhat.com>

tboerger committed:
feat: integrate all supported rocketchat attributes for alertmanagerconfig CR (#8016)

* feat: integrate all supported rocketchat attributes for alertmanagerconfig CR

* fix: drop rocketchat file references as requested

* fix: correctly format the generated code

* feat: move rocketchat validations into validation package

* chore: revert token/tokenID optional requirement

* feat: use ValidateTemplateURL for Rocket.Chat links

* Simplify code

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

---------

Signed-off-by: Simon Pasquier <spasquie@redhat.com>
Co-authored-by: Simon Pasquier <spasquie@redhat.com>

dependabot[bot] committed:
build(deps): bump imjasonh/setup-crane from 0.4 to 0.5

Bumps [imjasonh/setup-crane](https://github.com/imjasonh/setup-crane) from 0.4 to 0.5.
- [Release notes](https://github.com/imjasonh/setup-crane/releases)
- [Commits](https://github.com/imjasonh/setup-crane/compare/v0.4...v0.5)

---
updated-dependencies:
- dependency-name: imjasonh/setup-crane
  dependency-version: '0.5'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

kartikangiras committed:
add URL validation for SNS receiver (#8259)

* add url validation for sns receiver

* update validurl template

Signed-off-by: Kartik Angiras <angiraskartik@gmail.com>

---------

Signed-off-by: Kartik Angiras <angiraskartik@gmail.com>

simonpasquier committed:
chore: cut v0.89.0

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

simonpasquier committed:
chore: update go.mod files

Signed-off-by: Simon Pasquier <spasquie@redhat.com>

dependabot[bot] committed:
build(deps): bump golang.org/x/net from 0.49.0 to 0.50.0

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.49.0 to 0.50.0.
- [Commits](https://github.com/golang/net/compare/v0.49.0...v0.50.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-version: 0.50.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github.com/thanos-io/thanos from 0.40.1 to 0.41.0

Bumps [github.com/thanos-io/thanos](https://github.com/thanos-io/thanos) from 0.40.1 to 0.41.0.
- [Release notes](https://github.com/thanos-io/thanos/releases)
- [Changelog](https://github.com/thanos-io/thanos/blob/main/CHANGELOG.md)
- [Commits](https://github.com/thanos-io/thanos/compare/v0.40.1...v0.41.0)

---
updated-dependencies:
- dependency-name: github.com/thanos-io/thanos
  dependency-version: 0.41.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github.com/prometheus/alertmanager

Bumps [github.com/prometheus/alertmanager](https://github.com/prometheus/alertmanager) from 0.31.0 to 0.31.1.
- [Release notes](https://github.com/prometheus/alertmanager/releases)
- [Changelog](https://github.com/prometheus/alertmanager/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prometheus/alertmanager/compare/v0.31.0...v0.31.1)

---
updated-dependencies:
- dependency-name: github.com/prometheus/alertmanager
  dependency-version: 0.31.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

nutmos committed:
amcfg: update jira test case

remove duplicate vars

nutmos committed:
amcfg: update test cases

remove duplicates

nutmos committed:
amcfg:  move declare jira vars down

to be aligned with the previous commits

nutmos committed:
alertmanager: update generated files

nutmos commented:
Waiting for #8408 before updating the field.

nutmos committed:
alertmanager: fix merge conflicts

nutmos committed:
Merge branch 'main' into support-alertmanager-jira

nutmos committed:
alertmanager: update test cases

modernize pointer creation

nutmos committed:
alertmanager: update test case

fix linter

nutmos committed:
e2e: resolve merge conflicts

nutmos committed:
alertmanager: fix merge conflicts

nutmos committed:
alertmanager: update generated files

nutmos committed:
Merge branch 'main' into support-alertmanager-jira

nutmos committed:
Merge branch 'main' into support-alertmanager-jira

nutmos committed:
alertmanager: update jira validation

to match the new way of implementation

nutmos committed:
alertmanager: update error message

to match other receivers

nutmos committed:
alertmanager: update validation

to match newer format

nutmos committed:
alertmanager: add the CR test case

to ensure coverage

nutmos committed:
alertmanager: update api version

to match the test case

nutmos committed:
alertmanager: update test cases

remove unused variables

nutmos committed:
alertmanager: update test case

fix version

nutmos committed:
alertmanager: update test file

to match the test result

nutmos committed:
Merge branch 'main' into support-alertmanager-jira

nutmos committed:
alertmanager: fix merge conflicts

nutmos committed:
alertmanager: update generated files
